src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/IProjectSnapshotManager.cs

```diff
@@ -20,4 +22,19 @@ internal interface IProjectSnapshotManager
     bool TryGetLoadedProject(ProjectKey projectKey, [NotNullWhen(true)] out IProjectSnapshot? project);
 
     ImmutableArray<ProjectKey> GetAllProjectKeys(string projectFileName);
+
+    void Update(Action<ProjectSnapshotManager.Updater> updater);
+    void Update<TState>(Action<ProjectSnapshotManager.Updater, TState> updater, TState state);
+    TResult Update<TResult>(Func<ProjectSnapshotManager.Updater, TResult> updater);
+    TResult Update<TState, TResult>(Func<ProjectSnapshotManager.Updater, TState, TResult> updater, TState state);
```

Comment:
Normally I'm not a big fan of interfaces having lots of methods, where a reasonable portion of them could be written as extension methods. eg, one for sync, one for async, and wrappers that pass in empty state, convert actions in to funcs by returning null etc.

Though you've done the annoying work now, and there is unlikely to be another implementation ever, so not a huge deal at this point :)

Reply:
I definitely understand your feelings on this. I left it this way because it was easier to implement with all of the methods laid out in one place, and I didn't want to introduce any extra allocations with delegate wrappers. Wish we had default interface methods on .NET Framework. 😞